### ISSUE TITLE
Documentation (tl;dr -- Types are not documentation)

### ISSUE THREAD

**Issue by jasonzoladz** [paf31/purescript-thermite#15]
Thanks so much for creating Purescript and its book.  Both are truly fantastic.

I'm new to Purescript (with only 7 months of Haskell experience), so I'm not yet good enough to easily consume a Library like Thermite by following the types.  (read:  I've only recently grokked monad transformers)   It's possible, but it's a bit like reading hieroglyphics.  First I need to create a dictionary, then construct the rules of combination, and finally parse sentences.

As you know, documentation is a real problem in the Haskell community.  And good, detailed documentation isn't on the horizon.  However, we'd all benefit from the kind of minimal documentation provided by the Haskell prelude.  (e.g., see below)

I believe that you could greatly advance adoption of Purescript by setting a standard for what constitutes minimally complete documentation, and encouraging the community to follow the standard.

The "types are documentation" argument is ```Just lazy``` hipster-speak for "I couldn't be ```Eff```-ing bothered."
 
----------------------------------------------------------------------------------------

```maybe :: b -> (a -> b) -> Maybe a -> b Source```

The maybe function takes a default value, a function, and a Maybe value. If the Maybe value is Nothing, the function returns the default value. Otherwise, it applies the function to the value inside the Just and returns the result.

```data Either a b Source```

The Either type represents values with two possibilities: a value of type Either a b is either Left a or Right b.

The Either type is sometimes used to represent a value which is either correct or an error; by convention, the Left constructor is used to hold an error value and the Right constructor is used to hold a correct value (mnemonic: "right" also means "correct").

**Comment by jdegoes** [comment]
Fortunately, it's an open source project, and doc pull requests are very welcome. :smiley_cat:

**Comment by jdegoes** [comment]
Oh, by the way, we haven't announced it yet but you might be interested in [purescript-halogen](http://github.com/slamdata/purescript-halogen), which has docs already for most things.

**Comment by jasonzoladz** [comment]
I would be glad to create documentation for other folks' libraries.  (In fact, while I'm still getting my feet wet, it is perhaps the most responsible way for me to contribute to Purescript.)  But I'll need to spend an hour or so talking with the creator(s) so that I can understand the library.

The "it's an open source project" argument, however, is a cop out.  (It's used all over to justify half-assed work.)  I'm an attorney, so I can illustrate my point by analogy to pro-bono.

I was part of the team of attorneys that represented men of color in a class action against the NYPD for racial profiling.  (See http://www.huffingtonpost.com/robert-koehler/stop-and-frisk_b_3759231.html)  It wasn't work for pay but the professional standards were the same as for paying clients.  That is, do your best work.  (In fact, an attorney can be disciplined for failing to meet professional standards irrespective of compensation.)

No one requires you to draft open source libraries.  But that doesn't mean that you don't take on profession obligations when you release one.  A long way of saying:  do your best work. . .

I've seen purescript-halogen.  It looks promising.  But the kind of documentation you should aspire to is:  https://github.com/omcljs/om/wiki

Finally, purescript-aff is awesome!

**Comment by jasonzoladz** [comment]
Btw, I was also, in the early days of the suit, part of the team that represented IBM against SCO concerning IBM's contributions to the Linux kernel.  http://www.theregister.co.uk/2014/12/19/judge_spanks_sco_in_ancient_ownership_of_unix_lawsuit/
That is, I have much affection for open source.

**Comment by paf31** [comment]
Did you see the [video](https://www.youtube.com/watch?v=yrrOpjm-MzU) or the [todo-mvc example](https://github.com/paf31/purescript-thermite-todomvc)?

This library is still very experimental. I generally do write what I consider to be good docs, and I plan to write some for this when I think it is ready for public consumption, and when I have more time.

**Comment by jasonzoladz** [comment]
Phil,

Make no mistake, I am completely in awe of Purescript.

I submitted this as an issue re: Thermite, but it's not really directed at Thermite singularly.  Instead, it's a general suggestion about how you might shape the direction of a young language and it's community.  (I made the comment as an issue to a library you authored so as to avoid "making a scene" so to speak.  It's a minor criticism for a minor forum.)

Clojure and ClojureScript are killing it -- and it's not because Haskell/Purescript is too hard.  Have you seen the documentation of libraries in the Clojure/ClojureScript community?  It's fantastic.  It drives adoption; which drives the creation of new libraries . . .

Scala has Scalajs.  Clojure has ClojureScript.  Ocaml has js of ocaml.  Fairly mature solutions.  Haskell -- the elder -- has a web page about the Javascript problem.

I think Purescript is the solution.  But it needs a solid ecosystem.  I can't speak for others, but I'm not interested in Purescript because it's the new hotness or because it's niche.  (If that were the case, I'd be writing in Prof. Brady's Whitespace language.)  I really want type-safety in the browser.  

As the language author, I think you have the clout needed to foster good documentation (and perhaps discourage abandonware).

Anyway, thanks again for Purescript.

Cheers,
Jason

(Yes, I watched your Thermite presentation on Youtube -- three times.  :)  And while I do appreciate the effort, it perhaps suffers from being too close to the library.)
